onnx_xla/backend.cc

```diff
@@ -101,6 +101,28 @@ namespace onnx_xla {
     #undef OPERATION
   }
 
+  std::unique_ptr<Literal> XlaExecutor::descriptorToLiteral(const onnxTensorDescriptor& t) {
+    std::vector<int64> sizes;
+    for (auto i = 0; i < t.dimensions; ++i) {
+      sizes.push_back((int64) t.shape[i]);
+    }
+    int64 num_elements = std::accumulate(sizes.begin(), sizes.end(),
+                                         (int64) 1, std::multiplies<int64>());
+
+    #define OPERATION(type_from, type_to, vec)                                 \
+      auto l = std::unique_ptr<Literal>(new Literal(ShapeUtil::MakeShape(      \
+               NativeToPrimitiveType<type_to>(), sizes)));                     \
+      tensorflow::gtl::MutableArraySlice<type_to> l_data = l->data<type_to>(); \
+      type_from* inputData = (type_from*) t.buffer;                            \
+      for (auto i = 0; i < num_elements; ++i) {                                \
+        l_data[i] = (type_to) inputData[i];                                    \
+      }                                                                        \
+      return l;                                                                \
+
+    SWITCH(onnxifiToOnnx(t.dataType))
```

Comment:
This is very cryptic. Any way to avoid macros?

Reply:
Looks to me one macro for doing type switch is unavoidable, but we can use template function for `OPERATION` and let the type switch macro do the type dispatch.

```
#define dispatch(data_type, operation, ...) \
  swtich(data_type) {
  case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
    return operation<float, float>(__VA_ARGS__);
  }
  ...
  }

template<typename From, typename To>
std::unique_ptr<Literal> XlaExecutor::tensorToLiteralImpl(const Tensor& t) {
  // access the underlying data ptr via t.<From>data(), this needs to be added to the onnx Tensor class in onnx/common/tensor.h, similar to here https://github.com/pytorch/pytorch/blob/master/caffe2/core/tensor.h#L492-L506
  ...
}

std::unique_ptr<Literal> XlaExecutor::tensorToLiteral(const Tensor& t) {
  return dispatch(t.dataType, tensorToLiteralImpl, t);
}

template<typename From typename To>
std::unique_ptr<Literal> XlaExecutor::descriptorToLiteralImpl(const onnxTensorDescriptor& t) {
  ...
}

std::unique_ptr<Literal> XlaExecutor::descriptorToLiteral(const onnxTensorDescriptor& t) {
  return dispatch(t.dataType, descriptorToLiteralImpl, t);
}

}
```

functionality-wise it's same, just wanna make it more organized.

Reply:
Will refactor later in separate PR - once functionality added to onnx